### ISSUE TITLE
Prompt buffer VI formatting should instead call a mode-extensible generic

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#2219]
Currently we hard-code the VI formatting in the status formatter.  Instead, we should call a mode-specialized generic function, which VI-mode would then specialize.

Easy fix.

**Comment by Ambrevar** [comment]
On a similar topic, `print-status` should not be called manually.  It's too specific and too brittle.

Instead, a status-buffer should register a bunch of object to watch.  So if the status displays the list of modes, then it should `watch` for changes in the modes of the current buffer.

Is there a generic way to "watch for changes" in CL?
I cannot think of anything in the language or in a library, so what about this little "hack":

```lisp
(let ((method (defmethod (setf modes) :around (value (buffer modable-buffer))
                (declare (ignorable value buffer))
                (prog1 (call-next-method)
                  (print-status)))))
  (push method (watchers STATUS-BUFFER)))
```

In the above snippet, we stored the setf-around method so that it can be removed if need be.

The drawback of this approach is that it won't work if the setf-around method already exists.

Any other idea?

**Comment by Ambrevar** [comment]
Other drawback of the previous snippets: it only works on setf-able places.

**Comment by aadcg** [comment]
I agree, and noticed it today too. I'm just not sure if it's worth the effort just yet, see #2324.
